mac/Keyman4MacIM/Keyman4MacIM/TextApiCompliance.m

```diff
@@ -184,19 +192,85 @@ -(void) checkComplianceAfterInsert:(NSString *)insertedText deleted:(NSString *)
     return;
   }
   
+  // check whether the location has changed as expected
   NSRange selectionRange = [self.client selectedRange];
+  os_log_debug([KMLogs complianceLog], "  selectionRange: %{public}@", NSStringFromRange(selectionRange));
   if ([self validateNewLocation:selectionRange.location delete:deletedText]) {
     BOOL changeExpected = [self isLocationChangeExpectedOnInsert:deletedText insert:insertedText];
     BOOL locationChanged = [self hasLocationChanged:selectionRange];
     [self validateLocationChange:changeExpected hasLocationChanged:locationChanged];
   }
   
-  NSString *message = [NSString stringWithFormat:@"checkComplianceAfterInsert, self.hasWorkingSelectionApi = %@ for app %@", self.hasCompliantSelectionApi?@"YES":@"NO", self.clientApplicationId];
+  // only validate ability to read if we know we have ability to get selection
+  if (self.hasCompliantSelectionApi) {
+    [self validateCanReadInsertedText:insertedText location:selectionRange.location];
+  }
+  
+  NSString *message = [NSString stringWithFormat:@"checkComplianceAfterInsert, self.canReadText = %@, self.hasWorkingSelectionApi = %@ for app %@", self.canReadText?@"YES":@"NO", self.hasCompliantSelectionApi?@"YES":@"NO", self.clientApplicationId];
   os_log_debug([KMLogs complianceLog], "%{public}@", message);
   [KMSentryHelper addDebugBreadCrumb:@"compliance" message:message];
   [KMSentryHelper addApiComplianceTag:self.shortSentryTagDescription];
 }
 
+/**
+ * Confirm that we can get the string that ends at the current location,
+ * and the end of the string matches the string that  was just inserted
+ */
+- (void)validateCanReadInsertedText:(NSString *)insertedText location: (NSUInteger)newLocation {
+
+  NSString *context = [self getContext:newLocation];
+  
+  if (context && context.length > 0) {
+    if ([context hasSuffix:insertedText]) {
+      self.hasCompliantAttributedSubstringApi = YES;
+      os_log_debug([KMLogs complianceLog], "  validateCanReadInsertedText passed, context: '%{public}@', insertedText '%{public}@'", context, insertedText);
```

Comment:
If we don't want to log context in logs (KMInputMethodEventHandler.m:368), then is this logging also an issue? And several other places below.

Reply:
That's right -- I'll remove the related logs statements.